packages/@aws-cdk/aws-codepipeline/lib/action.ts

```diff
@@ -111,20 +226,272 @@ export interface ActionBindOptions {
 }
 
 export interface ActionConfig {
+  /**
+   * Static configuration returned by bind()
+   *
+   * @deprecated Return boundAction instead.
+   */
   readonly configuration?: any;
+
+  /**
+   * A bound version of the action
+   *
+   * Provides access to methods that can only be called once the
+   * Action has been bound to a pipeline.
+   *
+   * If this value is set, `configuration` will be ignored.
+   */
+  readonly boundAction?: IBoundAction;
+}
+
+/**
+ * The result of binding an Action
+ */
+export interface IBoundAction {
+  /**
+   * The unbound action
+   */
+  action: IAction;
+
+  /**
+   * Configuration for the Action
+   */
+  configuration(): any;
 }
 
 /**
  * A Pipeline Action
  */
 export interface IAction {
+  /**
+   * Information about the Action
+   */
   readonly actionProperties: ActionProperties;
 
+  /**
+   * Called when the Action is added to a Stage
+   */
   bind(scope: Construct, stage: IStage, options: ActionBindOptions): ActionConfig;
 
+  /**
+   * AWS Event that triggers when the Action changes state
+   */
   onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps): events.Rule;
 }
 
+/**
+ * Base class for CodePipeline Actions.
+ *
+ * @experimental
+ */
+export abstract class Action implements IAction {
+
+  private readonly _actionProperties: ActionProps;
+  private _pipeline?: IPipeline;
+  private _stage?: IStage;
+  private _scope?: Construct;
+  private readonly namespaceOrToken: string;
+  private actualNamespace?: string;
+  private variableReferenced = false;
+  private _actionName: string;
+  private _runOrder: number;
+  private _inputs: Artifact[];
+  private _outputs: Artifact[];
+
+  protected constructor(actionProperties: ActionProps) {
+    this._actionProperties = actionProperties;
+    this.namespaceOrToken = Lazy.stringValue({
+      produce: () => {
+        // make sure the action was bound (= added to a pipeline)
+        if (this.actualNamespace === undefined) {
+          throw new Error(`Cannot reference variables of action '${this._actionProperties.actionName}', ` +
+          'as that action was never added to a pipeline');
+        }
+
+        return this.finalNamespace;
+      },
+    });
+
+    this._inputs = actionProperties.inputs ?? [];
+    this._outputs = actionProperties.outputs ?? [];
+
+    this._actionName = actionProperties.actionName;
+    validation.validateName('Action', this._actionName);
+    validation.validateNamespaceName(actionProperties.variablesNamespace);
+
+    this._runOrder = actionProperties.runOrder ?? 1;
+    validateRunOrder(this._runOrder);
+  }
+
+  /**
+   * Return the properties for this action
+   *
+   * These properties are mutable.
+   */
+  public get actionProperties(): ActionProperties {
+    return {
+      ...this._actionProperties,
+      // Implied defaults here
+      owner: this._actionProperties.owner ?? 'AWS',
```

Comment:
This has to be incorrect, right? Why do we have the same defaulting in multiple places? I think there should be a single source of truths for the defaults, and it should be inside the CodePipeline construct, not any particular `IAction` implementation.

Reply:
I guess this is where we fundamentally disagree.

My position is: the `IAction` is ultimately responsible for rendering itself, and controlling how it appears in the pipeline. It can do that from scratch or inherit from `Action` in order to get a head start on a bunch of the implementation.

That also implies an Action is in charge of its own defaults, and is not going to answer `undefined` to a bunch of questions that could and show have definite answers.